backend/dns/dns_resolve.go

```diff
@@ -447,3 +439,81 @@ func LookupRDNS(ip net.IP, dnsServer string) (RecordType, error) {
 
 	return resp["TRACE"], nil
 }
+
+// LookupWithContext is like Lookup but supports context for timeout/cancellation and partial results.
+func LookupWithContext(ctx context.Context, dnsServer, recordType, hostname string, fullTrace bool) (Response, error) {
+	if !strings.HasSuffix(hostname, ".") {
+		hostname += "."
+	}
+	results := make(Response)
+	errCh := make(chan error, 1)
+	doneCh := make(chan struct{})
+
+	go func() {
+		var resp Response
+		var err error
+		if fullTrace {
+			resp, err = traceQueryWithContext(ctx, dnsServer, recordType, hostname)
+		} else {
+			resp, err = recursiveQueryWithContext(ctx, dnsServer, recordType, hostname)
+		}
+		if resp != nil {
+			for k, v := range resp {
+				results[k] = v
+			}
+		}
+		errCh <- err
+		close(doneCh)
+	}()
+
+	select {
+	case <-ctx.Done():
+		return results, ctx.Err()
+	case err := <-errCh:
+		return results, err
+	}
+}
+
+// traceQueryWithContext is like traceQuery but supports context for timeout/cancellation.
+func traceQueryWithContext(ctx context.Context, dnsServer, recordType, hostname string) (Response, error) {
+	respCh := make(chan Response, 1)
+	errCh := make(chan error, 1)
+	go func() {
+		resp, err := traceQuery(dnsServer, recordType, hostname)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		respCh <- resp
+	}()
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case err := <-errCh:
+		return nil, err
+	case resp := <-respCh:
+		return resp, nil
+	}
+}
+
+// recursiveQueryWithContext is like recursiveQuery but supports context for timeout/cancellation.
+func recursiveQueryWithContext(ctx context.Context, dnsServer, recordType, hostname string) (Response, error) {
+	respCh := make(chan Response, 1)
+	errCh := make(chan error, 1)
+	go func() {
+		resp, err := recursiveQuery(dnsServer, recordType, hostname)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		respCh <- resp
+	}()
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case err := <-errCh:
+		return nil, err
+	case resp := <-respCh:
+		return resp, nil
+	}
+}
```

Comment:
These wrappers only server to pass context down and cancel the underlying process after a timeout, however as discussed godns allows construction of dialer with many timeout attributes, we should instead make use of that.